src/env-inl.h

```diff
@@ -127,8 +127,8 @@ inline v8::Local<v8::String> Environment::AsyncHooks::provider_string(int idx) {
 
 inline void Environment::AsyncHooks::push_ids(double async_id,
                                               double trigger_id) {
-  CHECK_GE(async_id, 0);
-  CHECK_GE(trigger_id, 0);
+  CHECK_GE(async_id, -1);
```

Comment:
@AndreasMadsen 
>  My personal opinion is that those checks should simply be skipped when AsyncHooks isn't used.

It's not explicitly documented, but one goal of the API was to allow `async_hooks.executionAsyncId()` and `async_hooks.triggerAsyncId()` to be used regardless whether a hook has been enabled or not. We could possibly do the same trick that we do with domains and enable the checks if `async_hooks` is required, but domains is for performance reasons. Always performing these checks should be possible without ever causing the application to crash, and the additional cost probably isn't measurable.

> Relaxing the check when AsyncHooks is used, isn't a good solution.

I fully agree'd with this until I realized how badly node core can be abused by allowing users to pass in arbitrary objects instead of ones created by core constructors (e.g. timers). So it is necessary to coerce, at minimum, async ids of `undefined`.

> I really feel we are on a steap slope to some very difficult to debug errors.

Again, I fully agree'd with this until I came to terms with the reason pointed out above.

@refack 
> I tend to agree, but the -1 trick currently isn't public API, so we could change it if we agree on a better solution.

I was thinking about doing something more complex, like
* `-1` - indicates the `async_id_symbol` was `undefined`
* `-2` - indicates `async_id_symbol` was set but not a positive integer

etc. But figured that's too much to anticipate, and wouldn't be worth the time.